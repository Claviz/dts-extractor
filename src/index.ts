import fs from 'fs-extra';
import path from 'path';

export async function getDts({ nodeModulesPath, packages }: { nodeModulesPath: string; packages: string[] }) {
    const typings: { [key: string]: string } = {};
    const parsedPackages: { [key: string]: boolean } = {};

    async function getTypingsForPackages(packages: string[] = []) {
        for (const packageName of packages) {
            if (!parsedPackages[packageName]) {
                parsedPackages[packageName] = true;
                const packageJson = await fs.readJSON(`${nodeModulesPath}/${packageName}/package.json`, { throws: false });
                if (packageJson) {
                    const types = packageJson.typings || packageJson.types;
                    if (types) {
                        typings[`node_modules/${packageName}/package.json`] = JSON.stringify(packageJson);
                        const dirname = path.dirname(types) === '.' ? '' : path.dirname(types);
                        await getTypingsInDir(nodeModulesPath, `${packageName}${dirname ? '/' : ''}${dirname}`);
                    }
                    if (packageJson.dependencies) {
                        await getTypingsForPackages(Object.keys(packageJson.dependencies));
                    }
                }
            }
        }
    }

    async function getTypingsInDir(rootPath: string, path2: string) {
        const dts = (await fs.readdir(`${rootPath}/${path2}`));
        for (const fileName of dts) {
            if (fileName.endsWith('.d.ts')) {
                typings[`node_modules/${path2}/${fileName}`] = await fs.readFile(`${rootPath}/${path2}/${fileName}`, 'utf8');
            } else if ((await fs.lstat(`${rootPath}/${path2}/${fileName}`)).isDirectory()) {
                await getTypingsInDir(rootPath, `${path2}/${fileName}`);
            }
        }
    }

    await getTypingsForPackages(packages);

    return typings;
};