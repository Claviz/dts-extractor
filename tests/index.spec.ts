import { getDts } from '../src';

beforeEach(() => {
});

beforeAll(async () => {
});

afterAll(async () => {
});

beforeEach(async () => {
});

it('extracts type definition files', async () => {
    const result = await getDts({
        nodeModulesPath: './node_modules',
        packages: ['xlstream'],
    });

    expect(result).toMatchSnapshot();
});