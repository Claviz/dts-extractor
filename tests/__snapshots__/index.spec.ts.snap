// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extracts type definition files 1`] = `
Object {
  "node_modules/frac/package.json": "{\\"name\\":\\"frac\\",\\"types\\":\\"types\\"}",
  "node_modules/frac/types/index.d.ts": "/* frac.js (C) 2012-present SheetJS -- http://sheetjs.com */
// TypeScript Version: 2.2

export interface Frac$Module {
	(x: number, D: number, mixed?: boolean): [number, number, number];
	cont(x: number, D: number, mixed?: boolean): [number, number, number];
}
export const frac: Frac$Module;
export default frac;
",
  "node_modules/ssf/package.json": "{\\"name\\":\\"ssf\\",\\"types\\":\\"types\\"}",
  "node_modules/ssf/types/index.d.ts": "/* ssf.js (C) 2013-present SheetJS -- http://sheetjs.com */
// TypeScript Version: 2.2

/** Version string */
export const version: string;

/** Render value using format string or code */
export function format(fmt: string|number, val: any, opts?: any): string;

/** Load format string */
export function load(fmt: string, idx?: number): number;

/** Test if the format is a Date format */
export function is_date(fmt: string): boolean;


/** Format Table */
export interface SSF$Table {
	[key: number]: string;
	[key: string]: string;
}

/** Get format table */
export function get_table(): SSF$Table;

/** Set format table */
export function load_table(tbl: SSF$Table): void;


/** Parsed date */
export interface SSF$Date {
	/** number of whole days since relevant epoch, 0 <= D */
	D: number;
	/** integral year portion, epoch_year <= y */
	y: number;
	/** integral month portion, 1 <= m <= 12 */
	m: number;
	/** integral day portion, subject to gregorian YMD constraints */
	d: number;
	/** integral day of week (0=Sunday .. 6=Saturday) 0 <= q <= 6 */
	q: number;

	/** number of seconds since midnight, 0 <= T < 86400 */
	T: number;
	/** integral number of hours since midnight, 0 <= H < 24 */
	H: number;
	/** integral number of minutes since the last hour, 0 <= M < 60 */
	M: number;
	/** integral number of seconds since the last minute, 0 <= S < 60 */
	S: number;
	/** sub-second part of time, 0 <= u < 1 */
	u: number;
}

/** Parse numeric date code */
export function parse_date_code(v: number, opts?: any): SSF$Date;
",
  "node_modules/xlstream/lib/index.d.ts": "/// <reference types=\\"node\\" />
import { Transform } from 'stream';
import { IXlsxStreamOptions, IWorksheetOptions } from './types';
export declare function getXlsxStream(options: IXlsxStreamOptions): Promise<Transform>;
export declare function getWorksheets(options: IWorksheetOptions): Promise<string[]>;
",
  "node_modules/xlstream/lib/types.d.ts": "export interface IXlsxStreamOptions {
    filePath: string;
    sheet: string | number;
    withHeader?: boolean;
    ignoreEmpty?: boolean;
}
export interface IWorksheetOptions {
    filePath: string;
}
",
  "node_modules/xlstream/package.json": "{\\"name\\":\\"xlstream\\",\\"types\\":\\"lib/index\\"}",
}
`;
